packages/table/src/Table.tsx

```diff
@@ -35,19 +77,79 @@ export default function Table<Shape>({
   className,
   ...rest
 }: TableProps<Shape>) {
+  const [scrollState, setScrollState] = React.useState<ScrollState>(
+    ScrollState.None,
+  );
+  const divRef = React.useRef<HTMLDivElement>(null);
+  const viewportSize = useViewportSize();
+
+  useIsomorphicLayoutEffect(() => {
+    const ref = divRef.current;
+
+    if (ref == null) {
+      return;
+    }
+
+    if (ref.scrollWidth > ref.clientWidth) {
+      setScrollState(ScrollState.Right);
+    } else if (
+      viewportSize != null &&
+      ref.getBoundingClientRect().width <= viewportSize.width
+    ) {
+      setScrollState(ScrollState.None);
+    }
+  }, [viewportSize]);
+
+  const handleScroll = (e: React.UIEvent & { target: HTMLElement }) => {
+    const { scrollWidth, clientWidth: elementWidth } = e.target;
+    const isScrollable = scrollWidth > elementWidth;
+
+    if (isScrollable) {
+      const scrollPosition = e.target.scrollLeft;
+      const maxPosition = scrollWidth - elementWidth;
+
+      if (scrollPosition > 0 && scrollPosition < maxPosition) {
+        setScrollState(ScrollState.Both);
+      } else if (scrollPosition > 0) {
+        setScrollState(ScrollState.Left);
+      } else if (scrollPosition < maxPosition) {
+        setScrollState(ScrollState.Right);
+      }
+    }
+  };
+
+  const debounceScroll = debounce(handleScroll, 50, { leading: true });
+
+  const onScroll: React.UIEventHandler<HTMLDivElement | HTMLPreElement> = e => {
```

Comment:
kinda gross, but i think we need something like
```suggestion
  const onScroll: React.EventHandler<React.UIEvent & { target: HTMLDivElement | HTMLPreElement }> = e => {
```

Reply:
mm, it was a valiant effort but looks like React needs to fix the typing of `onScroll` for our types to work. type assertions it is then 😔

Reply:
Reverted back 😞